# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/webscheduling/WebTaskSchedulerMainThread.cpp
# Commit: 4ccb6970887e
# Full Hash: 4ccb6970887ece212a1176681a567480ea586dfc
# Author: Sean Feng <sefeng@mozilla.com>
# Date: 2025-02-25 09:30:56
# Regressor Bug: 1920115
# File Overlap Count: 2
# Description:
#   Bug 1920115 - Allow high priority tasks to run before timers r=smaug
#   
#   In Gecko, unlike other runnables where the TaskController
#   picks one each at a time, the TimeoutManager allows multiple
#   timers to run sequentially without allowing other runnables
# ==============================================================================

diff -r 7522970c5a1d -r 4ccb6970887e dom/webscheduling/WebTaskSchedulerMainThread.cpp
--- a/dom/webscheduling/WebTaskSchedulerMainThread.cpp	Mon Feb 24 19:31:28 2025 +0000
+++ b/dom/webscheduling/WebTaskSchedulerMainThread.cpp	Mon Feb 24 19:31:28 2025 +0000
@@ -12,6 +12,8 @@
 
 namespace mozilla::dom {
 
+uint32_t gNumNormalOrHighPriorityQueuesHaveTaskScheduledMainThread = 0;
+
 NS_IMETHODIMP WebTaskMainThreadRunnable::Run() {
   if (mScheduler) {
     RefPtr<WebTask> task = mScheduler->GetNextTask(true /* aIsMainThread */);
@@ -52,4 +54,16 @@
       NS_DispatchToMainThreadQueue(runnable.forget(), aPriority));
   return true;
 }
+
+void WebTaskSchedulerMainThread::
+    IncreaseNumNormalOrHighPriorityQueuesHaveTaskScheduled() {
+  ++gNumNormalOrHighPriorityQueuesHaveTaskScheduledMainThread;
+}
+
+void WebTaskSchedulerMainThread::
+    DecreaseNumNormalOrHighPriorityQueuesHaveTaskScheduled() {
+  MOZ_DIAGNOSTIC_ASSERT(
+      gNumNormalOrHighPriorityQueuesHaveTaskScheduledMainThread > 0);
+  --gNumNormalOrHighPriorityQueuesHaveTaskScheduledMainThread;
+}
 }  // namespace mozilla::dom